# json_to_db.py
from pathlib import Path
import json
from clinvar_query.utils.paths import database_file, clinvar_folder
from clinvar_query.modules.insert_annotated_results import insert_clinvar, insert_patient_information, insert_variants

def json_to_dir():
    # Directory containing JSON files generated by clinvar_api_query
    json_dir = Path(clinvar_folder)
    json_files = list(json_dir.glob("*.json"))
    

    if not json_files:
        print(f"⚠️ No JSON files found in {json_dir}")
    else:
        print(f"Found {len(json_files)} JSON files in {json_dir}")

    for json_file in json_files:
        print(f"Processing file: {json_file.name}")
        try:
            with open(json_file, "r") as f:
                variants_data = json.load(f)
                p_id = Path(json_file).stem
                patient_id = p_id.split("_")[0]
        except Exception as e:
            print(f"❌ Failed to load {json_file}: {e}")
            continue

        for entry in variants_data:
            variant_str = entry.get("variant")
            g_hgvs = entry.get("g_hgvs")
            clinvar_ids = entry.get("clinvar_ids", [])
            summary = entry.get("esummary", {})

            if not variant_str or not g_hgvs:
                print(f"⚠️ Skipping entry, missing variant or g_hgvs: {entry}")
                continue

            # Extract key fields from esummary if available
            chromosome = summary.get(g_hgvs, {}).get("chromosome")
            gene = summary.get(g_hgvs, {}).get("gene_symbol")
            classification = summary.get(g_hgvs, {}).get("germline_classification")
            star_rating = summary.get(g_hgvs, {}).get("review_status")
            allele_frequency = summary.get(g_hgvs, {}).get("value")
            associated_conditions = summary.get(g_hgvs, {}).get("trait_name")

            clinvar = {
                "variant_id" : variant_str,
                "consensus_classification": classification,
                "hgvs": g_hgvs,
                "associated_conditions": associated_conditions,
                "gene": gene,
                "star_rating": star_rating,
                "allele_frequency": allele_frequency,
                "chromosome": chromosome
            }
            patient_info = {
                "patient_id" : patient_id,
                "id_test_type" : p_id
            }
            variants = {
                "variant_id" : variant_str,
                "id_test_type": p_id,
                "patient_variant": p_id + variant_str,
            }
            insert_patient_information(patient_info)
            insert_variants(variants)
            insert_clinvar(clinvar)


    print("✅ All ClinVar JSON files processed into the database.")


if __name__ == "__main__":
    json_to_dir()